include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+#ifndef GNURADIO_DATA_SINK_HPP
+#define GNURADIO_DATA_SINK_HPP
+
+#include "circular_buffer.hpp"
+#include "dataset.hpp"
+#include "history_buffer.hpp"
+#include "node.hpp"
+#include "tag.hpp"
+
+#include <any>
+#include <chrono>
+
+namespace fair::graph {
+
+enum class blocking_mode { NonBlocking, Blocking };
+
+enum class trigger_observer_state {
+    Start,        ///< Start a new dataset
+    Stop,         ///< Finish dataset
+    StopAndStart, ///< Finish pending dataset, start a new one
+    Ignore        ///< Ignore tag
+};
+
+// TODO is the scope where want these?
+struct null_type {};
+
+// Until clang-format can handle concepts
+// clang-format off
+template<typename T>
+concept TriggerPredicate = requires(const T p, tag_t tag) {
+    { p(tag) } -> std::convertible_to<bool>;
+};
+
+template<typename T>
+concept TriggerObserver = requires(T o, tag_t tag) {
+    { o(tag) } -> std::convertible_to<trigger_observer_state>;
+};
+
+template<typename T>
+concept TriggerObserverFactory = requires(T f) {
+    { f() } -> TriggerObserver;
+};
+// clang-format on
+
+template<typename T>
+class data_sink;
+
+struct data_sink_query {
+    std::optional<std::string> sink_name;
+    std::optional<std::string> signal_name;
+
+    static data_sink_query
+    with_signal_name(std::string_view name) {
+        return { {}, std::string{ name } };
+    }
+
+    static data_sink_query
+    with_sink_name(std::string_view name) {
+        return { std::string{ name }, {} };
+    }
+};
+
+class data_sink_registry {
+    std::mutex            _mutex;
+    std::vector<std::any> _sinks;
+
+public:
+    // TODO this shouldn't be a singleton but associated with the flow graph (?)
+    static data_sink_registry &
+    instance() {
+        static data_sink_registry s_instance;
+        return s_instance;
+    }
+
+    template<typename T>
+    void
+    register_sink(data_sink<T> *sink) {
+        std::lock_guard lg{ _mutex };
+        _sinks.push_back(sink);
+    }
+
+    template<typename T>
+    void
+    unregister_sink(data_sink<T> *sink) {
+        std::lock_guard lg{ _mutex };
+        std::erase_if(_sinks, [sink](const std::any &v) {
+            try {
+                return std::any_cast<data_sink<T> *>(v) == sink;
+            } catch (...) {
+                return false;
+            }
+        });
+    }
+
+    template<typename T>
+    std::shared_ptr<typename data_sink<T>::poller>
+    get_streaming_poller(const data_sink_query &query, blocking_mode block = blocking_mode::NonBlocking) {
```

Comment:
default should be 'blocking'

Reply:
From our discussions I got that most users would use non-blocking? Can change this of course.

Reply:
Nope... most users want/need a data delivery guarantee... Handling dropped samples is and advanced use-case for those who know what they are doing.